sklearn/preprocessing/tests/test_imputation.py

```diff
@@ -358,3 +358,101 @@ def test_imputation_copy():
+    n_imputed_features = len(imputer_with_in.imputed_features_)
+    assert_array_equal(imputer_with_in.imputed_features_, np.array([0, 2, 3]))
+    assert_equal(Xt_with_in.shape,
+                 (n_samples, n_features_new + n_imputed_features))
+
+    imputer = Imputer(missing_values=-1, strategy='mean', axis=1)
+    imputer_with_in = Imputer(missing_values=-1, strategy='mean', axis=1,
+                              add_missing_indicator=True)
+    Xt = imputer.fit_transform(X)
+    Xt_with_in = imputer_with_in.fit_transform(X)
+    imputed_features_mask = X[:, imputer_with_in.imputed_features_] == -1
+    assert_array_equal(Xt_with_in, np.hstack((Xt, imputed_features_mask)))
+    n_features_new = Xt.shape[1]
+    n_imputed_features = len(imputer_with_in.imputed_features_)
+    assert_array_equal(imputer_with_in.imputed_features_, np.array([0, 1, 2, 3]))
+    assert_equal(Xt_with_in.shape,
+                 (n_samples, n_features_new + n_imputed_features))
+
+    # one feature with all missing values and one with no missing value
+    # when axis=0 the feature gets discarded
+    X = np.array([
+       [-1,  -1,   1,   3],
+       [4,  -1,   1,  -1],
+       [8,  -1,   1,  11],
+       [12,  -1,   1,  15],
+       [16,  -1,   1,  19]
+    ])
+    imputer = Imputer(missing_values=-1, strategy='mean', axis=0)
+    imputer_with_in = Imputer(missing_values=-1, strategy='mean', axis=0,
+                              add_missing_indicator=True)
+    Xt = imputer.fit_transform(X)
+    Xt_with_in = imputer_with_in.fit_transform(X)
+    imputed_features_mask = X[:, imputer_with_in.imputed_features_] == -1
+    assert_array_equal(Xt_with_in, np.hstack((Xt, imputed_features_mask)))
+    n_features_new = Xt.shape[1]
+    n_imputed_features = len(imputer_with_in.imputed_features_)
+    assert_array_equal(imputer_with_in.imputed_features_, np.array([0, 3]))
+    assert_equal(Xt_with_in.shape,
+                 (n_samples, n_features_new + n_imputed_features))
+
+    imputer = Imputer(missing_values=-1, strategy='mean', axis=1)
+    imputer_with_in = Imputer(missing_values=-1, strategy='mean', axis=1,
+                              add_missing_indicator=True)
+    Xt = imputer.fit_transform(X)
+    Xt_with_in = imputer_with_in.fit_transform(X)
+    imputed_features_mask = X[:, imputer_with_in.imputed_features_] == -1
+    assert_array_equal(Xt_with_in, np.hstack((Xt, imputed_features_mask)))
+    n_features_new = Xt.shape[1]
+    n_imputed_features = len(imputer_with_in.imputed_features_)
+    assert_array_equal(imputer_with_in.imputed_features_, np.array([0, 1, 3]))
+    assert_equal(Xt_with_in.shape,
+                 (n_samples, n_features_new + n_imputed_features))
+
+def test_missing_indicator_sparse():
+    # one feature with all missng values
+    X = sparse.csc_matrix([
+       [-1,  -1,   2,   3],
+       [4,  -1,   6,  -1],
+       [8,  -1,  10,  11],
+       [12,  -1,  -1,  15],
+       [16,  -1,  18,  19]
+    ])
```

Comment:
Yeah I was more focussed on getting the sparse case to work. Sorry. But I suppose you meant make more entries as missing since the missing value isn't zero in this branch of code we are testing.